openedx_events/learning/data.py

```diff
@@ -0,0 +1,88 @@
+"""
+Data attributes for events within the architecture subdomain `learning`.
+
+These attributes follow the form of attr objects specified in OEP-49 data
+pattern.
+"""
+from typing import Dict
+
+import attr
+from opaque_keys.edx.keys import CourseKey
+
+
+@attr.s(frozen=True)
+class UserProfileData:
+    """
+    Attributes defined for Open edX student's profile object.
+    """
+
+    meta = attr.ib(type=Dict[str, str], factory=dict)
+    name = attr.ib(type=str, factory=str)
+
+
+@attr.s(frozen=True)
+class StudentData:
+    """
+    Attributes defined for Open edX student object.
+    """
+
+    username = attr.ib(type=str)
+    email = attr.ib(type=str)
+    first_name = attr.ib(type=str, factory=str)
+    last_name = attr.ib(type=str, factory=str)
+    is_active = attr.ib(type=bool, default=True)
+    profile = attr.ib(type=UserProfileData, default=UserProfileData())
+
+
+@attr.s(frozen=True)
+class RegistrationData:
+    """
+    Attributes defined for Open edX registration profile object.
+    """
+
+    activation_key = attr.ib(type=str)
+
+
+@attr.s(frozen=True)
+class CourseOverviewData:
```

Comment:
Is the course overview what we want to pass around? or is it the course data directly?

Reply:
We will probably fill the data in via what is in the CourseOverview data to avoid querying mongo, but still CourseOverview seems to me like something that is too lms implementation specific

Reply:
Right. I'll change it to CourseData.

Besides `course_key` and `display_name`, do you have any other course fields in mind? 

These are all we can access without querying mongo i.e from CourseOverview model:

    # Course identification
    id = CourseKeyField(db_index=True, primary_key=True, max_length=255)
    _location = UsageKeyField(max_length=255)
    org = TextField(max_length=255, default='outdated_entry') **--> SELECTED but is outdated?**
    display_name = TextField(null=True)
    display_number_with_default = TextField()
    display_org_with_default = TextField()

    start = DateTimeField(null=True) **--> SELECTED**
    end = DateTimeField(null=True) **--> SELECTED**

    # These are deprecated and unused, but cannot be dropped via simple migration due to the size of the downstream
    # history table. See DENG-19 for details.
    # Please use start and end above for these values.
    start_date = DateTimeField(null=True)
    end_date = DateTimeField(null=True)

    advertised_start = TextField(null=True)
    announcement = DateTimeField(null=True)

    # URLs
    # Not allowing null per django convention; not sure why many TextFields in this model do allow null
    banner_image_url = TextField()
    course_image_url = TextField() **--> SELECTED (can be use in advertisement?) **
    social_sharing_url = TextField(null=True) **--> SELECTED (can be use in advertisement?) **
    end_of_course_survey_url = TextField(null=True)

    # Certification data
    certificates_display_behavior = TextField(null=True)
    certificates_show_before_end = BooleanField(default=False)
    cert_html_view_enabled = BooleanField(default=False)
    has_any_active_web_certificate = BooleanField(default=False)
    cert_name_short = TextField()
    cert_name_long = TextField()
    certificate_available_date = DateTimeField(default=None, null=True)  

    # Grading
    lowest_passing_grade = DecimalField(max_digits=5, decimal_places=2, null=True) **--> SELECTED**

    # Access parameters
    days_early_for_beta = FloatField(null=True)
    mobile_available = BooleanField(default=False)
    visible_to_staff_only = BooleanField(default=False)
    _pre_requisite_courses_json = TextField()  # JSON representation of list of CourseKey strings

    # Enrollment details
    enrollment_start = DateTimeField(null=True)
    enrollment_end = DateTimeField(null=True)
    enrollment_domain = TextField(null=True)
    invitation_only = BooleanField(default=False)
    max_student_enrollments_allowed = IntegerField(null=True)

    # Catalog information
    catalog_visibility = TextField(null=True)
    short_description = TextField(null=True)
    course_video_url = TextField(null=True)
    effort = TextField(null=True)
    self_paced = BooleanField(default=False)
    marketing_url = TextField(null=True)
    eligible_for_financial_aid = BooleanField(default=True)

    # Course highlight info, used to guide course update emails
    has_highlights = NullBooleanField(default=None)  # if None, you have to look up the answer yourself

    language = TextField(null=True)

    history = HistoricalRecords()